src/Components/Server/src/Circuits/CircuitRegistry.cs

```diff
@@ -0,0 +1,140 @@
+            ILogger<CircuitRegistry> logger)
+        {
+            _options = options.Value;
+            _logger = logger;
+
+            ActiveCircuits = new ConcurrentDictionary<string, CircuitHost>(StringComparer.Ordinal);
+
+            InactiveCircuits = new MemoryCache(new MemoryCacheOptions
+            {
+                SizeLimit = _options.MaxRetainedDisconnectedCircuits,
+            });
+
+            _postEvictionCallback = new PostEvictionCallbackRegistration
+            {
+                EvictionCallback = OnEntryEvicted,
+            };
+        }
+
+        internal ConcurrentDictionary<string, CircuitHost> ActiveCircuits { get; }
+
+        internal MemoryCache InactiveCircuits { get; }
+
+        public void Register(CircuitHost circuitHost)
+        {
+            ActiveCircuits.TryAdd(circuitHost.CircuitId, circuitHost);
+        }
+
+        public void Deactivate(CircuitHost circuitHost, string connectionId)
+        {
+            if (!ActiveCircuits.TryGetValue(circuitHost.CircuitId, out circuitHost))
+            {
+                // The circuit might already have been marked as inactive.
+                return;
+            }
+
+            if (!string.Equals(circuitHost.CircuitClient.ConnectionId, connectionId, StringComparison.Ordinal))
+            {
+                // The circuit is associated with a different connection. One way this could happen is when
+                // the client reconnects with a new connection before the OnDisconnect for the older
+                // connection is executed. Do nothing
+                return;
+            }
+
+            ActiveCircuits.TryRemove(circuitHost.CircuitId, out circuitHost);
+
+            circuitHost.CircuitClient.Connected = false;
+            var entryOptions = new MemoryCacheEntryOptions
+            {
+                AbsoluteExpiration = DateTimeOffset.UtcNow.Add(_options.DisconnectedCircuitRetentionPeriod),
+                Size = 1,
+                PostEvictionCallbacks = { _postEvictionCallback },
+            };
+
+            InactiveCircuits.Set(circuitHost.CircuitId, circuitHost, entryOptions);
+        }
+
+        public bool TryActivate(string circuitId, IClientProxy clientProxy, string connectionId, out CircuitHost host)
+        {
+            if (ActiveCircuits.TryGetValue(circuitId, out host))
+            {
+                if (!string.Equals(host.CircuitClient.ConnectionId, connectionId, StringComparison.Ordinal))
+                {
+                    // The host is still active i.e. the server hasn't detected the client disconnect.
+                    // However the client reconnected establishing a new connection.
+                    host.CircuitClient.Transfer(clientProxy, connectionId);
+                }
+
+                return true;
+            }
+
+            if (InactiveCircuits.TryGetValue(circuitId, out host))
+            {
+                ActiveCircuits.TryAdd(circuitId, host);
```

Comment:
There's another race here between 93-95 - two thread coulds both *claim* the circuit in between these two lines.